src/server/pps/server/pipeline_controller.go

```diff
@@ -145,19 +146,35 @@ func (a *apiServer) step(pachClient *client.APIClient, pipeline string, keyVer,
 			if err := op.scaleUpPipeline(); err != nil {
 				return err
 			}
-			return op.setPipelineState(pps.PipelineState_PIPELINE_RUNNING)
+			return op.setPipelineState(pps.PipelineState_PIPELINE_RUNNING, "")
 		}
 		// don't want cron commits or STANDBY state changes while pipeline is
 		// stopped
 		op.stopPipelineMonitor()
 		// default: scale down if pause/standby hasn't propagated to etcd yet
 		return op.scaleDownPipeline()
 	case pps.PipelineState_PIPELINE_FAILURE:
-		// pipeline fails if docker image isn't found
+		// pipeline fails if it encounters an unrecoverable error
 		if err := op.finishPipelineOutputCommits(); err != nil {
 			return err
 		}
 		return op.deletePipelineResources()
+	case pps.PipelineState_PIPELINE_CRASHING:
+		if !op.rcIsFresh() {
+			return op.restartPipeline("stale RC") // step() will be called again after etcd write
+		}
+		if op.pipelineInfo.Stopped {
+			return op.setPipelineState(pps.PipelineState_PIPELINE_PAUSED, "")
+		}
+		workersUp, err := a.allWorkersUp(pachClient, op.pipelineInfo)
+		if err != nil {
+			return err
+		}
+		if workersUp {
+			return op.setPipelineState(pps.PipelineState_PIPELINE_RUNNING, "")
+		}
+		time.Sleep(10 * time.Second)
```

Comment:
I'm not totally sure about this line, it prevents the controller from busy looping when a pipeline is crashing, but I'm not sure if there's a better place to do it. At the very least I probably shouldn't be using a hard coded value here.

Reply:
I think I see how this works. One alternative that occurs to me would be to spawn a goroutine, similar to `monitorPipeline`, that watches the workers and handles the transition to `RUNNING`. Instead of looping in the controller, the controller is unblocked and the loop happens separately off to the side (and the side loop is responsible for putting the pipeline into `RUNNING`, like `monitorPipeline`).

I think this would at least makes the sleep less obtrusive (and it could eventually be replaced with kubernetes watches, I think) and it removes the need to set the pipeline state to `CRASHING` which gets rid of the controller feedback loop.